### ISSUE TITLE
diversity function strange method of NA assignment

### ISSUE THREAD

**Issue by dmcglinn** [vegandevs/vegan#239]
I noticed this line of code for the function `diversity`

https://github.com/vegandevs/vegan/blob/master/R/diversity.R#L29
```
if (any(is.na(total)))
    is.na(H) <- is.na(total)
```
This effectively replaces values in the vector `H` with NAs if the corresponding cell in the vector `total` is NA. I think a clearer way to make this replacement would be `H[is.na(total)] = NA` the if statement is not needed at this point either.

**Comment by jarioksa** [comment]
The `if` statement is there for reason: nothing will be done if there are no `NA`s.

Your code and the current **vegan** code are equivalent and give the same results. However, it really seems that using `H[is.na(total)]` is faster than using `is.na<-.default()` to set the `NA` values. It seems that your suggestion copies `H` once and the current **vegan** code copies `H` twice.

I'll see this: I'll probably keep the `if()`, but will change setting `NA`.

**Comment by jarioksa** [comment]
See issue #187

**Comment by jarioksa** [comment]
We use the `any(is.na(total))` to save time. We assume that normally we do not have `NA` in the community data and don't do any replacement. The `if()` takes care that we don't use time for doing nothing if nothing needs be done. Replacing `H` with itself takes time because we need to copy `H` before we write over `H`:

```R
H <- runif(1000, 1, 100)
tracemem(H)
# [1] "<0x7fc369e034b0>"
H[] <- H
# tracemem[0x7fc369e034b0 -> 0x7fc36cb31b80]:
```

However, the equivalent `is.na<-` we use now makes two copies:

```R
is.na(H) <- is.na(H)
# tracemem[0x7fc36cb33b70 -> 0x7fc369cfb760]: 
# tracemem[0x7fc369cfb760 -> 0x7fc369cfb1f0]: is.na<-.default is.na<- 
```

This is also visible in `microbenchmark::microbenchmark()` where the current `is.na<-` code is much slower than the `H[] <-` you suggested, and skipping any copying with `if(any(is.na()))`  is still 2x faster than doing `H[] <- H` in any case.
```
Unit: nanoseconds
                     expr  min     lq    mean median     uq   max neval
                   H <- H  143  160.5  170.67  169.0  177.0   247   100
                 H[] <- H 1261 1384.0 1924.42 1496.0 1665.0 14958   100
    H[is.na(total)] <- NA 1177 1351.5 1812.47 1495.5 1717.5 11332   100
 is.na(H) <- is.na(total) 3629 3966.0 5334.35 4183.0 5681.0 71675   100
            any(is.na(H))  653  704.0  943.62  752.0  871.5  7002   100
```

**Comment by gavinsimpson** [comment]
@jarioksa are you proposing replacing the code @dmcglinn mentioned with:

```r
if (any(NAS <- is.na(total)))
    H[NAS] <- NA
```
? If so, that seems fine to me.

**Comment by jarioksa** [comment]
@gavinsimpson , that exactly was on my mind (although `microbenchmark` found no advantage of `NAS <- is.na(total)` assignment).

**Comment by jarioksa** [comment]
Closed via 522a20a649
